packages/test/src/test-typed-search-attributes.ts

```diff
@@ -0,0 +1,437 @@
+    // Transform untyped keys into 'untypedKey: IndexValueType' pairs.
+    const untypedKeys = Object.entries(untypedAttrsInput).reduce(
+      (acc, [key, value]) => {
+        const typedKey = TypedSearchAttributes.getKeyFromUntyped(key, value);
+        const encodedKey = encodeSearchAttributeIndexedValueType(typedKey?.type);
+        if (encodedKey) {
+          acc[key] = encodedKey;
+        }
+        return acc;
+      },
+      {} as { [key: string]: temporal.api.enums.v1.IndexedValueType }
+    );
+
+    const typedKeys = typedAttrsListInput.reduce(
+      (acc, pair) => {
+        const encodedKey = encodeSearchAttributeIndexedValueType(pair.key.type);
+        if (encodedKey) {
+          acc[pair.key.name] = encodedKey;
+        }
+        return acc;
+      },
+      {} as { [key: string]: temporal.api.enums.v1.IndexedValueType }
+    );
+
+    await t.context.env.connection.operatorService.addSearchAttributes({
+      namespace: t.context.env.namespace,
+      searchAttributes: {
+        ...untypedKeys,
+        ...typedKeys,
+        ...erroneousTypedKeys,
+      },
+    });
+
+    await waitUntil(async () => {
+      const resp = await t.context.env.connection.operatorService.listSearchAttributes({
+        namespace: t.context.env.namespace,
+      });
+      return (
+        Object.keys(untypedKeys).every((key) => key in resp.customAttributes) &&
+        Object.keys(typedKeys).every((key) => key in resp.customAttributes)
+      );
+    }, 300);
+  });
+}
+
+test('does not allow non-integer values for integer search attributes', async (t) => {
+  try {
+    const { taskQueue } = helpers(t);
+    const client = t.context.env.client;
+    const action: ScheduleOptionsAction = {
+      type: 'startWorkflow',
+      workflowType: dummyWorkflow,
+      taskQueue,
+    };
+    const erroneousKeyName = Object.keys(erroneousTypedKeys)[0];
+    await client.schedule.create({
+      scheduleId: randomUUID(),
+      spec: {
+        calendars: [{ hour: { start: 2, end: 7, step: 1 } }],
+      },
+      action,
+      typedSearchAttributes: [
+        // Use a double value for an integer search attribute.
+        // This is legal at compile-time, but should error at runtime when converting to payload.
+        { key: { name: erroneousKeyName, type: SearchAttributeType.INT }, value: 123.4 },
+      ],
+    });
+  } catch (err) {
+    if (err instanceof Error) {
+      t.is(err.message, 'Invalid search attribute value 123.4 for given type INT');
+    } else {
+      t.fail('Unexpected error type');
+    }
+  }
+});
+
+interface TestInputSearchAttributes {
+  input: {
+    searchAttributes?: SearchAttributes;
+    typedSearchAttributes?: TypedSearchAttributes | SearchAttributePair[];
+  };
+  expected: {
+    searchAttributes?: SearchAttributes;
+    typedSearchAttributes?: TypedSearchAttributes;
+  };
+}
+
+// inputTestCases contains permutations of search attribute inputs
+const inputTestCases: TestInputSearchAttributes[] = [
```

Comment:
nit: When doing this kind of dataset-based tests, it's generally useful to also include a "name" or "comment" on each entry, and output it in the assertion. Makes it much easier to when tests fail.

In your assertion below, you would do somehting like:

```
inputTestCases.map(async ({ input, expected, name }) => {
  t.deepEqual(desc.searchAttributes, expected.searchAttributes, name)
}
```

Reply:
Definitely useful, added.